### PR TITLE
Filter variant tranches removes old filters by default

### PR DESCRIPTION
**Pull request by lucidtronix**
This addresses: https://github.com/broadinstitute/gatk/issues/4928
@cmnbroad care to review?

### CODE REVIEW

File: src/main/java/org/broadinstitute/hellbender/tools/walkers/vqsr/FilterVariantTranches.java [hunk 1]
```diff
@@ -82,6 +82,9 @@
     @Argument(fullName = "info-key", shortName = "info-key", doc = "The key must be in the INFO field of the input VCF.")
     private String infoKey = GATKVCFConstants.CNN_2D_KEY;
 
+    @Argument(fullName = "keep-old-filters", shortName = "keep-old-filters", doc = "Keeps filters already in the VCF.", optional=true)
```
[the remark this answers is not in the document]
Comment (lucidtronix) [reply src/main/java/org/broadinstitute/hellbender/tools/walkers/vqsr/FilterVariantTranches.java]: My motivation for adding this was getting burned by a VCF which had been filtered by VQSR, so I definitely want the ability to remove filters added by other tools, even if it is not the default. My vote is for it to be the default and to emit a warning for every filter removed from the header. @ldgauthier thoughts?

**Comment by lucidtronix** [comment]
screenshot 2018-08-01 11 42 49

After seeing plots like the one above, where we have lots of false negative SNPs and lots of false positive indels, it became clear that we should have separate tranche sensitivites for the two variant types.  I've added that into this PR. I made the filter removal opt-in as suggested and it now shares the argument with VariantFiltration.

@cmnbroad back to you, sorry for moving the goal line for this PR.

**Comment by cmnbroad** [comment]
@lucidtronix Just got back to this and noticed that many of the tests failed. Some look like transient git-lfs issues (so I restarted them), but [this one](https://travis-ci.org/broadinstitute/gatk/jobs/410870690) looks real. Can you take a look before I re-review ?

**Comment by lucidtronix** [comment]
@cmnbroad you're right, I forgot to update the CNNVariantPipeline test with the new arguments.  Should be fixed now.

File: src/main/java/org/broadinstitute/hellbender/tools/walkers/vqsr/FilterVariantTranches.java [hunk 2]
```diff
@@ -133,8 +139,8 @@ public void firstPassApply(final VariantContext variant, final ReadsContext read
 
     @Override
     public void afterFirstPass() {
-        logger.info(String.format("Found %d SNPs %d indels with INFO score key:%s.", scoredSnps, scoredIndels, infoKey));
-        logger.info(String.format("Found %d SNPs %d indels in the resources.", snpScores.size(), indelScores.size()));
+        logger.info(String.format("Found %d SNPs and %d indels with INFO score key:%s.", scoredSnps, scoredIndels, infoKey));
+        logger.info(String.format("Found %d SNPs and %d indels in the resources.", snpScores.size(), indelScores.size()));
```
Comment (cmnbroad) [inline src/main/java/org/broadinstitute/hellbender/tools/walkers/vqsr/FilterVariantTranches.java]: For clarity can you rename snpScores/indelScores to something like resourceSNPScores/resourceIndelScores or truthSNPScores/truthIndelScores.

File: src/main/java/org/broadinstitute/hellbender/tools/walkers/vqsr/FilterVariantTranches.java [hunk 3]
```diff
@@ -183,47 +194,66 @@ public void closeTool() {
...
+                return filterKeyFromTranches(type, infoKey, tranches.get(i-1), tranches.get(i));
             }
         }
-        return filterKeyFromTranches(infoKey, tranches.get(tranches.size()-1), 100.0);
+        return filterKeyFromTranches(type, infoKey, tranches.get(tranches.size()-1), 100.0);
+    }
+
+    private List<Double> validateTranches(List<Double> tranches){
+        if (tranches.size() < 1 || tranches.stream().anyMatch(d -> d < 0 || d >= 100.0)){
+            throw new GATKException("At least 1 tranche value must be given and all tranches must be greater than 0 and less than 100.");
```
Comment (cmnbroad) [inline src/main/java/org/broadinstitute/hellbender/tools/walkers/vqsr/FilterVariantTranches.java]: This should throw CommandLineException.

File: src/main/java/org/broadinstitute/hellbender/tools/walkers/vqsr/FilterVariantTranches.java [hunk 4]
```diff
@@ -133,8 +139,8 @@ public void firstPassApply(final VariantContext variant, final ReadsContext read
 
     @Override
     public void afterFirstPass() {
-        logger.info(String.format("Found %d SNPs %d indels with INFO score key:%s.", scoredSnps, scoredIndels, infoKey));
-        logger.info(String.format("Found %d SNPs %d indels in the resources.", snpScores.size(), indelScores.size()));
+        logger.info(String.format("Found %d SNPs and %d indels with INFO score key:%s.", scoredSnps, scoredIndels, infoKey));
+        logger.info(String.format("Found %d SNPs and %d indels in the resources.", snpScores.size(), indelScores.size()));
 
         if (scoredSnps == 0 || scoredIndels == 0 || snpScores.size() == 0 || indelScores.size() == 0){
             throw new GATKException("VCF must contain SNPs and indels with scores and resources must contain matching SNPs and indels.");
```
Comment (cmnbroad) [inline src/main/java/org/broadinstitute/hellbender/tools/walkers/vqsr/FilterVariantTranches.java]: This should be a UserException. GATKException is for unexpected errors in the code that are outside the user’s control, like the one thrown by filterStringFromScore.

File: src/main/java/org/broadinstitute/hellbender/tools/walkers/filters/VariantFiltration.java [hunk 5]
```diff
@@ -170,7 +169,7 @@
     /**
      * Invalidate previous filters applied to the VariantContext, applying only the filters here
      */
-    @Argument(fullName=INVALIDATE_LONG_NAME,doc="Remove previous filters applied to the VCF",optional=true)
+    @Argument(fullName=StandardArgumentDefinitions.REMOVE_OLD_FILTERS_LONG_NAME, doc="Remove previous filters applied to the VCF", optional=true)
```
[the remark this answers is not in the document]
Comment (lucidtronix) [reply src/main/java/org/broadinstitute/hellbender/tools/walkers/filters/VariantFiltration.java]: I liked my pithy argument, but fair point, fixed.

File: src/main/java/org/broadinstitute/hellbender/tools/walkers/vqsr/FilterVariantTranches.java [hunk 6]
```diff
@@ -82,6 +88,9 @@
     @Argument(fullName = "info-key", shortName = "info-key", doc = "The key must be in the INFO field of the input VCF.")
     private String infoKey = GATKVCFConstants.CNN_2D_KEY;
 
+    @Argument(fullName = StandardArgumentDefinitions.REMOVE_OLD_FILTERS_LONG_NAME, doc = "Remove filters already in the VCF.", optional=true)
```
[the remark this answers is not in the document]
Comment (lucidtronix) [reply src/main/java/org/broadinstitute/hellbender/tools/walkers/vqsr/FilterVariantTranches.java]: fixed

File: src/main/java/org/broadinstitute/hellbender/tools/walkers/vqsr/FilterVariantTranches.java [hunk 7]
```diff
@@ -183,47 +194,66 @@ public void closeTool() {
...
+        hInfo.addAll(inputHeader.getMetaDataInSortedOrder());
 
-        if( tranches.size() >= 2 ) {
-            for(int i = 0; i < tranches.size() - 1; i++) {
-                String filterKey = filterKeyFromTranches(infoKey, tranches.get(i), tranches.get(i+1));
-                String filterDescription = filterDescriptionFromTranches(infoKey, tranches.get(i), tranches.get(i+1));
-                hInfo.add(new VCFFilterHeaderLine(filterKey, filterDescription));
-            }
+        if (removeOldFilters){
+            hInfo.removeIf(x -> x instanceof VCFFilterHeaderLine);
```
[the remark this answers is not in the document]
Comment (lucidtronix) [reply src/main/java/org/broadinstitute/hellbender/tools/walkers/vqsr/FilterVariantTranches.java]: fixed

File: src/main/java/org/broadinstitute/hellbender/tools/walkers/vqsr/FilterVariantTranches.java [hunk 8]
```diff
@@ -67,11 +67,17 @@
...
+    @Argument(fullName="snp-tranche",
+            shortName="snp-tranche",
             doc="The levels of truth sensitivity at which to slice the data. (in percents, i.e. 99.9 for 99.9 percent and 1.0 for 1 percent)",
             optional=true)
-    private List<Double> tranches = new ArrayList<>(Arrays.asList(99.9, 99.99));
+    private List<Double> snpTranches = new ArrayList<>(Arrays.asList(99.9, 99.99));
+
+    @Argument(fullName="indel-tranche",
+            shortName="indel-tranche",
+            doc="The levels of truth sensitivity at which to slice the data. (in percents, i.e. 99.9 for 99.9 percent and 1.0 for 1 percent)",
```
[the remark this answers is not in the document]
Comment (lucidtronix) [reply src/main/java/org/broadinstitute/hellbender/tools/walkers/vqsr/FilterVariantTranches.java]: fixed

File: src/test/java/org/broadinstitute/hellbender/tools/walkers/vqsr/FilterVariantTranchesIntegrationTest.java [hunk 9]
```diff
@@ -48,8 +50,11 @@ public void testTrancheFilteringDuplicate() throws IOException {
                 .addArgument(StandardArgumentDefinitions.OUTPUT_LONG_NAME, "%s")
                 .addArgument("resource", snpTruthVCF)
                 .addArgument("resource", indelTruthVCF)
-                .addArgument("tranche", "99.0")
-                .addArgument("tranche", "99.0")
+                .addArgument("snp-tranche", "99.5")
```
[the remark this answers is not in the document]
Comment (lucidtronix) [reply src/test/java/org/broadinstitute/hellbender/tools/walkers/vqsr/FilterVariantTranchesIntegrationTest.java]: fixed

File: src/test/java/org/broadinstitute/hellbender/tools/walkers/vqsr/FilterVariantTranchesIntegrationTest.java [hunk 10]
```diff
@@ -101,4 +114,27 @@ public void testTrancheFilteringTranchesOrder() throws IOException {
...
+
+        final ArgumentsBuilder argsBuilder = new ArgumentsBuilder();
+        argsBuilder.addArgument(StandardArgumentDefinitions.VARIANT_LONG_NAME, trancheVCF)
+                .addArgument(StandardArgumentDefinitions.OUTPUT_LONG_NAME, "%s")
+                .addArgument("resource", snpTruthVCF)
+                .addArgument("resource", indelTruthVCF)
+                .addArgument("snp-tranche", "99.0")
+                .addArgument("indel-tranche", "99.0")
+                .addArgument("info-key", "MIX_SMALL_2D_W_DROPOUT")
+                .addArgument("remove-old-filters", "false")
```
[the remark this answers is not in the document]
Comment (lucidtronix) [reply src/test/java/org/broadinstitute/hellbender/tools/walkers/vqsr/FilterVariantTranchesIntegrationTest.java]: fixed

File: src/main/java/org/broadinstitute/hellbender/tools/walkers/vqsr/FilterVariantTranches.java (hunk 2 above)
Comment (lucidtronix) [reply src/main/java/org/broadinstitute/hellbender/tools/walkers/vqsr/FilterVariantTranches.java]: fixed

File: src/main/java/org/broadinstitute/hellbender/tools/walkers/vqsr/FilterVariantTranches.java (hunk 3 above)
Comment (lucidtronix) [reply src/main/java/org/broadinstitute/hellbender/tools/walkers/vqsr/FilterVariantTranches.java]: fixed

File: src/main/java/org/broadinstitute/hellbender/tools/walkers/vqsr/FilterVariantTranches.java (hunk 4 above)
Comment (lucidtronix) [reply src/main/java/org/broadinstitute/hellbender/tools/walkers/vqsr/FilterVariantTranches.java]: fixed

**Comment by lucidtronix** [comment]
@cmnbroad Thanks for the review, back to you!

File: src/main/java/org/broadinstitute/hellbender/tools/walkers/vqsr/FilterVariantTranches.java [hunk 11]
```diff
@@ -133,36 +148,41 @@ public void firstPassApply(final VariantContext variant, final ReadsContext read
...
         final VariantContextBuilder builder = new VariantContextBuilder(variant);
-
+        if (removeOldFilters) {
+            builder.unfiltered();
+        }
         if (variant.hasAttribute(infoKey)) {
             final double score = Double.parseDouble((String) variant.getAttribute(infoKey));
             if (variant.isSNP() && isTrancheFiltered(score, snpCutoffs)) {
-                builder.filter(filterStringFromScore(score, snpCutoffs));
+                builder.filter(filterStringFromScore("SNP", score, snpTranches, snpCutoffs));
```
Comment (cmnbroad) [inline src/main/java/org/broadinstitute/hellbender/tools/walkers/vqsr/FilterVariantTranches.java]: Extract string constants for "SNP" and "INDEL" and use them in the calls to `filterStringFromScore ` and `addTrancheHeaderFields'.

**Comment by cmnbroad** [comment]
Back to @lucidtronix.

**Comment by cmnbroad** [comment]
@lucidtronix I also just noticed that the tool-level gatkdoc for this tool is incorrect. It has a line that says: 

> Apply tranche filters based on CNN_1D scores

File: src/main/java/org/broadinstitute/hellbender/tools/walkers/vqsr/FilterVariantTranches.java (hunk 11 above)
Comment (lucidtronix) [reply src/main/java/org/broadinstitute/hellbender/tools/walkers/vqsr/FilterVariantTranches.java]: done

**Pull request merged**